### PR TITLE
Disable tunnel for http2 requests

### PR DESCRIPTION
**Pull request by Nightrider0098**
Closes https://github.com/urllib3/urllib3/issues/3298

### CODE REVIEW

**Review by pquentin — COMMENTED** [review]
Thanks, can you please add an unit test to make sure this isn't lost in a later refactoring?

Also, please make sure to add "Closes #XXXX" in your pull request description. This closes the issue automatically when the pull request is merged, and helps understanding what you're fixing. I've just edited your pull request to include that.

**Comment by sethmlarson** [comment]
I'm also imagining a test case that checks the bytes sent over the socket for the absence of the "h2" ALPN in the TLS ClientHello, I believe we do something similar for server name?

**Comment by pquentin** [comment]
> I'm also imagining a test case that checks the bytes sent over the socket for the absence of the "h2" ALPN in the TLS ClientHello, I believe we do something similar for server name?

I did not know, but we do have a SNI socketlevel test: `test_hostname_in_first_request_packet`. And we even have an ALPN test: https://github.com/urllib3/urllib3/blob/d7bb83b48c23d8dd429c20a43ac3da74ea6e9df0/test/with_dummyserver/test_socketlevel.py#L108-L133

We will eventually have to adapt this to test to include checking that the `h2` ALPN is included for normal requests, but that can happen in another PR. I also tried checking what happens when setting up a tunnel in a modified version of this test, but I realized that this pull request makes sure the request does not even go out (raises `NotImplementedError`).

And since the request does not go out, it does not have to be a test that includes a server in `test_with_dummyserver/test_proxy_poolmanager.py`. It can be in `test/test_proxymanager.py`. When using the `http_version` fixture and `http_version` is equal to `http2`, doing a [tunnel request (with HTTP CONNECT)](https://urllib3.readthedocs.io/en/stable/advanced-usage.html#proxies) should raise `NotImplementedError`.

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,4 @@
 /build
 /docs/_build
 coverage.xml
+venv/
```
Comment (sigmavirus24) [inline .gitignore:11]: I'd prefer we not add things here. You should use a global gitignore if there's a way you're working that is different from the project conventions

**Review by sigmavirus24 — CHANGES_REQUESTED** [review]

**Comment by Nightrider0098** [comment]
@pquentin Should i override set_tunnel in connection pool or writing the check in __init__ is fine?

**Review by sethmlarson — COMMENTED** [review]
Thank you for this PR, I have a suggestion that'll ensure the test case is useful into the near future where `HTTPSConnection` begins speaking HTTP/2 on its own (without a separate connection).

The approach would be:

* Setup a socket server that records the handshake bytes
* Send a tunnel proxy connection with TLS handshake
* Assert that the bytes for an ALPN of `http/1.1` are there, but not `h2`.

This way in the future when the two connections are combined together this test case is still useful and it removes the need for the other tests.

**Comment by sethmlarson** [comment]
See this test which verifies that the hostname is in the initial TLS handshake via SNI: https://github.com/urllib3/urllib3/blob/main/test/with_dummyserver/test_socketlevel.py#L103

Can assert the `http/1.1` and `h2` ALPN strings in a similar way.

**Comment by pquentin** [comment]
Thanks! Can you please look into the CI failures?

File: test/with_dummyserver/test_socketlevel.py [hunk 2]
```diff
@@ -1133,6 +1133,34 @@ def echo_socket_handler(listener: socket.socket) -> None:
...
+                    "Content-Type: text/plain\r\n"
+                    "Content-Length: %d\r\n"
+                    "\r\n"
+                    "%s" % (len(self.buf), self.buf.decode("utf-8"))
+                ).encode("utf-8")
+            )
+            sock.close()
+
+        self._start_server(echo_socket_handler)
+        base_url = f"http://{self.host}:{self.port}"
```
Comment (sethmlarson) [inline test/with_dummyserver/test_socketlevel.py:1157]: This is using HTTPS, tunneling and ALPN are TLS features so requires HTTPS. It's expected for the handshake to fail, we want to see what ALPN values are being offered and not complete the handshake.

File: test/with_dummyserver/test_socketlevel.py [hunk 3]
```diff
@@ -1133,6 +1133,34 @@ def echo_socket_handler(listener: socket.socket) -> None:
...
+        with proxy_from_url(base_url) as proxy:
+            r = proxy.request("GET", "http://google.com/")
+            assert r.status == 200
+            assert b"HTTP/1.1" in self.buf
```
Comment (sethmlarson) [inline test/with_dummyserver/test_socketlevel.py:1161]: For ALPN the value is `b"http/1.1"`.

**Review by sethmlarson — CHANGES_REQUESTED** [review]

**Comment by pquentin** [comment]
> Thank you for this PR, I have a suggestion that'll ensure the test case is useful into the near future where `HTTPSConnection` begins speaking HTTP/2 on its own (without a separate connection). [...]

I'm confused by this message. Are you suggesting that enabling HTTP CONNECT to an otherwise HTTP/2 capable host should downgrade to HTTP/1.1? I also don't understand what "speaking HTTP/2 on its own" and "without a separate connection" mean. I think laying out what should happen now and later and how this test will cover both cases would help me a lot.

**Comment by sethmlarson** [comment]
@pquentin Sorry that comment is indeed a bit confusing!

> `HTTPSConnection` begins speaking HTTP/2 on its own (without a separate connection). [...]

This was meant to mean when `urllib3.connection.HTTPSConnection` automatically uses HTTP/2, instead of through `urllib3.http2.connection.HTTP2Connection`.

File: test/with_dummyserver/test_connectionpool.py [hunk 4]
```diff
@@ -452,7 +452,6 @@ def test_socket_timeout_updated_on_reuse_parameter(
             )
 
     def test_tunnel(self, http_version: str) -> None:
-        # note the actual httplib.py has no tests for this functionality
```
Comment (pquentin) [inline test/with_dummyserver/test_connectionpool.py:455]: This has been tested for 10 years now! 5 years after urllib3's first commit. https://github.com/python/cpython/blob/e001027188001f4bdf6ea16f70726ca0fabe85c4/Lib/test/test_httplib.py#L2225

**Review by sethmlarson — APPROVED** [review]
Thanks Quentin!

**Comment by pquentin** [comment]
@Nightrider0098 Thanks for your help here! While I brought this to the finish line, the $100 bounty associated to this issue is yours and you can claim it on OpenCollective.

**Pull request merged**